patch taken from 
http://lists.gnu.org/archive/html/bug-texinfo/2007-10/msg00020.html
(for cmds.c) and adjusted for insertion.c by Norbert Preining
---
 makeinfo/cmds.c      |   10 +++++++---
 makeinfo/insertion.c |    2 +-
 2 files changed, 8 insertions(+), 4 deletions(-)

Index: texinfo-4.11.dfsg.1/makeinfo/cmds.c
===================================================================
--- texinfo-4.11.dfsg.1.orig/makeinfo/cmds.c	2007-07-15 19:36:19.000000000 +0200
+++ texinfo-4.11.dfsg.1/makeinfo/cmds.c	2007-12-19 16:19:24.000000000 +0100
@@ -1637,10 +1637,14 @@
 
       if (!no_headers && !html)
         {
+          /* use add_* instead of insert_* because otherwise the
+             file header ("This is ...") will end up inside the
+             dir section markers.  */
           kill_self_indent (-1); /* make sure there's no indentation */
-          insert_string ("INFO-DIR-SECTION ");
-          insert_string (line);
-          insert ('\n');
+          cm_noindent (); /* make sure again */
+          add_word ("INFO-DIR-SECTION ");
+          add_word (line);
+          add_char ('\n');
         }
 
       free (line);
Index: texinfo-4.11.dfsg.1/makeinfo/insertion.c
===================================================================
--- texinfo-4.11.dfsg.1.orig/makeinfo/insertion.c	2007-12-19 16:19:35.000000000 +0100
+++ texinfo-4.11.dfsg.1/makeinfo/insertion.c	2007-12-19 16:19:41.000000000 +0100
@@ -488,7 +488,7 @@
       close_single_paragraph ();
       filling_enabled = no_indent = 0;
       inhibit_paragraph_indentation = 1;
-      insert_string ("START-INFO-DIR-ENTRY\n");
+      add_word ("START-INFO-DIR-ENTRY\n");
       break;
 
     case documentdescription:
